### ISSUE TITLE
[doc] Setting up PerlIO callbacks when embedding a Perl interpreter using C (before any `*.pm` modules were ever loaded or any Perl code executed)

### ISSUE THREAD

**Issue by vadimkantorov** [Perl/perl5#22571]
Hi!

I managed to do a fully hermetic single-file static build of perl via building all modules statically (followed https://perldoc.perl.org/perlembed) and providing my own implementations of open/fopen/read/seek to serve `*.pm` system files from memory.

Is there a way to hook up to the Perl's own PerlIO layers system to make sure that Perl only calls these functions (including for module/`*.pm` discovery and loading) and never goes to libc's IO functions or does libc IO function calls / IO syscalls? This would be much cleaner and a more robust solution.

It would be nice if setting up PerlIO in perlembed scenario was covered in docs.

I also wonder how diamond operator is implemented in the code and which functions from https://github.com/Perl/perl5/blob/blead/perlio.c it calls and in what sequence (e.g. for `perl -e 'open(f,"<","my.txt");print(<f>);'` and for `perl -e 'open(f,"<","my.txt");$line=<f>;print($line);'`)

Thanks!

---

If anyone's curious to see what my hack looks like - https://github.com/vadimkantorov/perlpack, but it's very much a WIP

My current problem is that overriding `open` /`close` / `read` / `stat` / `lseek` / `access` / `fopen` / `fileno` was sufficient for `perl -e 'use Cwd;print(Cwd::cwd(),"\n");'`, so it can successfully discover and load the `Cwd.pm` file from my virtual read-only FS, but doing `perl -e 'open(F,"<","/mnt/perlpack/.../Cwd.pm");print(<F>);'` does not work - probably because Perl is trying to do fcntl/ioctl/some other version of stat call and I am not implementing these. In any case, it is currently not invoking the `read` function for some reason when I'm using the diamond operator because of some failures on the way. Which IO/stdio calls are used by Perl in a typical opening/reading a file? `strace` shows `open` -> `fcntl` -> `ioctl` -> `lseek` -> `fstat` -> `mmap` -> `read`, but these are raw syscalls, so I'm wondering what are the concrete libc/stdio IO functions (I imagine this is somewhere in `perlio.c` or `do_io.c` but there are quite a few of indirection layers - so hard to parse through by a novice in the perl's codebase) are used by Perl in a typical opening/reading a file (e.g. stat has many variants) - so that I can override them.

**Comment by tonycoz** [comment]
There is PERL_IMPLICIT_SYS, but that replaces *all* I/O (not just module loading) and only has a host implementation on Windows.

If you just want modules to be loaded from memory you can add a hook to `@INC` that checks for a known name and loads that module from memory, see perldoc -f require.

**Comment by vadimkantorov** [comment]
I'll check out `PERL_IMPLICIT_SYS` - replacing all I/O is fine for my usecase, as my custom I/O functions only serve from in-memory for some special prefixes like `/mnt/perl`. Is anywhere any docs / examples of using `PERL_IMPLICIT_SYS` to override? (and what functions need to be overridden for ensuring both module loading and for `perl -e 'open(f,"<","my.txt");print(<f>);'`. I'm only concerned for compiling/running on Linux for now.

> If you just want modules to be loaded from memory you can add a hook to `@INC` that checks for a known name and loads that module from memory, see perldoc -f require.

Actually, interested both for modules and for regular, basic file reads. For modules, can such INC-hook be added via C perlembed interface (without executing Perl code)?

> and only has a host implementation on Windows. 

And regarding PerlIO infra, is it relevant for my usecase (module / `*.pm` loads and regular basic file reads)? Can it be configured via C perlembed interface? Or would you recommend using `PERL_IMPLICIT_SYS`? Or is using `PERL_IMPLICIT_SYS` on Linux impossible?

Thank you!

**Comment by Leont** [comment]
> I'll check out PERL_IMPLICIT_SYS - replacing all I/O is fine for my usecase, as my custom I/O functions only serve from in-memory for some special prefixes like /mnt/perl. Is anywhere any docs / examples of using PERL_IMPLICIT_SYS to override? (and what functions need to be overridden for ensuring both module loading and for perl -e 'open(f,"<","my.txt");print(<f>);'. I'm only concerned for compiling/running on Linux for now.

It's only ever been done before for Windows but there's no reason it would be impossible on Linux. See `perlhost.h`, `win32.c` and `perllib.c` in `win32/` for prior art.

**Comment by vadimkantorov** [comment]
Thanks for the pointers! I'll look into what entails using `PERL_IMPLICIT_SYS` on Linux.

And maybe the last question, if you would know if PerlIO can also be used for this I/O override goal? And if so, can it be configured via a C API before any Perl code gets executed?

**Comment by tonycoz** [comment]
> For modules, can such INC-hook be added via C perlembed interface (without executing Perl code)?

After perl_construct() something like:
```
CV *hook = newXS("MyPackage::my_hook", \&xs_my_hook_xs, __FILE__);
AV *inc = get_av("INC");
av_unshift(inc, 1);
av_store(inc, 0, newRV_noinc(hook));
```
You could also define the hook sub in perl with eval_pv()/eval_sv().

> And maybe the last question, if you would know if PerlIO can also be used for this I/O override goal? And if so, can it be configured via a C API before any Perl code gets executed?

You might be able to do it by modifying `PL_def_layerlist` or via PERLIO in the environment, but I've never tried it.

It also won't allow you to hook operations like stat() and fcntl().

**Comment by vadimkantorov** [comment]
> You might be able to do it by modifying `PL_def_layerlist` or via PERLIO in the environment, but I've never tried it.

This would be interesting if this worked, and it could do for just `*.pm` embeds into the binary! It seems that `stat()` is not used on the `*.pm` search / loading path...

Basically, I was looking for a portable way to have hooks to embed and have loaded `*.pm` in fully static https://perldoc.perl.org/perlembed setup... I achieved this using static musl libc symbol renaming (https://github.com/vadimkantorov/perlpack), but having this using PerlIO hooks would be a nicer and a more portable way.

**Comment by vadimkantorov** [comment]
> `CV *hook = newXS("MyPackage::my_hook", \&xs_my_hook_xs, __FILE__);`

Would be grateful if anyone could suggest an example how this `void xs_my_hook_xs(pTHX_ CV* cv) {}` C-land INC hook should look like. If I understand it should pop from `cv` the reference to itself and the requested file name an then somehow push a return value?

From https://perldoc.perl.org/functions/require:
```
INC hook will be called with two parameters, the first a reference to itself, and the second the name of the file to be included (e.g., Foo/Bar.pm). The subroutine should return either nothing or else a list of up to four values in the following order:

A reference to a scalar, containing any initial source code to prepend to the file or generator output.

A filehandle, from which the file will be read.

A reference to a subroutine. If there is no filehandle (previous item), then this subroutine is expected to generate one line of source code per call, writing the line into [$_](https://perldoc.perl.org/perlvar#%24_) and returning 1, then finally at end of file returning 0. If there is a filehandle, then the subroutine will be called to act as a simple source filter, with the line as read in [$_](https://perldoc.perl.org/perlvar#%24_). Again, return 1 for each valid line, and 0 after all lines have been returned. For historical reasons the subroutine will receive a meaningless argument (in fact always the numeric value zero) as $_[0].

Optional state for the subroutine. The state is passed in as $_[1].
```

**Comment by Leont** [comment]
I don't think overriding PerlIO is the way to go here. That is, unless you want the modules to be open()able as if they really existed.

Pushing your own handler (either a coderef or an object) into `@INC` is probably the best approach, in either case you need to add some code to core to handle that. An alternative approach might be to override the require opcode.

**Comment by vadimkantorov** [comment]
Yeah, I think figuring out both the avenue of installing a INC-hook from C land (for embeding modueles' pm/pl files) and the PerlIO avenue (for embedding data-files installed into the Perl prefix) would be very useful nice for making basic embedded, static Perls more practical, simple and transparent.

It also would allow more simple compilation of Perl to WASI target (which might not have the built-in virtual FS feature as opposed to Emscripten)

**Comment by Leont** [comment]
> It also would allow more simple compilation of Perl to WASI target (which might not have the built-in virtual FS feature as opposed to Emscripten)

That really sounds like you should want to go the implicit sys route. That way you can virtualize all system/io interactions.

**Comment by Leont** [comment]
It probably would help a lot if we'd first write a generic implicit sys implementation, because you probably don't want to redefine all of it.

**Comment by vadimkantorov** [comment]
Yeah, I think ideally examples of both would be very useful:
1. Generic PerlIO impls with example of override from C land
2. INC hook installation from C land (for simpler and more specific cases of emdedding only pm/pl module files)

**Comment by vadimkantorov** [comment]
Regarding avenue (2) (with INC hook): another way could be populating from C land a path=>modulesourcecode dictionary in the Perl interpreter state, and then adding the INC hook written in Perl which would access this dictionary. How can one do that nicely without having to prepare a perl statement like `"mymodulesources["..."] = "..."` (this is problematic because we need to escape the string, so ideally we'd create the string var from C land directly and then directly push it to the `mymodulesources` dictionary)?

Could I use somehow `HeSVKEY_set` (from https://perldoc.perl.org/perlapi) for this end? Just need to create a string var as `SV` somehow... Maybe via `newSVpv`?

**Comment by vadimkantorov** [comment]
As a full example, here is what I am currently doing to override the `*.pm` (and some data files) loads from embedded prefix tree: https://github.com/busytex/busytex/blob/988f0a3337f461ab5cada8d410581d449edc032b/.github/workflows/build-biber.yml#L175-L669 for the https://github.com/plk/biber instead of using PAR / shared library linking

Defining custom IO functions as PerlIO layer / PERL_IMPLICIT_SYS would be more explicit and robust than using the `--wrap` switch of the linker and rolling some custom fd impls etc

**Comment by bulk88** [comment]
> > It also would allow more simple compilation of Perl to WASI target (which might not have the built-in virtual FS feature as opposed to Emscripten)
> 
> That really sounds like you should want to go the implicit sys route. That way you can virtualize all system/io interactions.

Yeah, the design purpose of iperlsys.h is for runtime dynamic swapping of which LibC disk file/impl/vendor the perl interp /libperl is connected to or using. Perl on Symbian Mobile OS was the HIGHEST consumer of iperlsys.h' s / PERL_IMPLICIT_SYS 'S api, 2nd highest was Perl on Netware, 3rd highest was Win32 perl. Since # 1 and # 2 ports were deprecated and deleted from CORE long long ago, that leaves Win32 Perl as the # 1 and only consumer of that API.

IIRC Symbian Perl had an even nastier OS arch design problem it had to work around than Win32 Perls problems with fork. Symbian Mobile OS did not have a concept of  a "process" or a PID number in its public API. All Symbian "Apps" / "executable binaries" were actually same process same address space "plugins", think .so/.dll files. 

The real life impact of that means, all C static global data vars and extern C global data vars, were shared memory between ALL perl "OS Processes. All root apps (.exe ???)s and all .so/.dll files  were mmaped only once  on Symbian OS and then stayed mmaped in shared inter-process memory , for almost eternity. I think when they last perl.bin/perl.exe/perl.jar "process" exited , Symbian OS would demap perl.bin/libperl.so, and therefore terminate the lifespan of libperl's C global vars.

But luck trying to use the perl language without ever shell starting a 2nd perl process to exist in parallel with the first process (EUMM/cpan.pl/Test::More require 2 perl process es to exist at 1 moment in time).

Internally, Symbian OS shared each binary's RW .data and .bss sections, between processes just like how read only .text and .rdata/.rodata are shared  between processes.  Symbian didn't use COW on .data/.bss sections like Linux and Windows do. A write to .data section in 1 process would appear in all other processes that mapped that same binary.

ON Win32 perl, ActiveState devs in the mid 1990s were expecting Larry Wall to lock the official  P5P repository in the near future, and terminate Perl 5 exactly like Larry terminated the Perl 4 platform. Perl 5 would've been terminated because a production grade "Perl 6" was expected to be released in 18 months or less back then. 

So ActiveState designed iperlsys.h/PERL_IMPLICIT_SYS to be an API where they (ActiveState) can release new and improved Perl 5 engines/interps/distros without any participation or "authorization" from Perl 5 Porters people.

The concept is similar to ancient and modern
 ECMAScript 5 interps, and the ECMAScript 5.0 Specification. A standalone build of the V8 or Spidermonkey interps, will have ZERO communication with the outside world. No console, no I/O , no sockets, no file descriptors, nor any other syscall type methods are mentioned in the ES/JS language spec. Perhaps the Date() class in Javascript can be called the "I" part of I/O if you stretch the truth.

I belive that was the purpose of iperlsys.h, P5P maintains an ascii src code parser/bytecode/optree compiler, and 3rd parties maintain all other "base class library" features bundled with a standalone /bin/perl binary or a /bin/perl + it's bundled /lib/*.pm files.

That would be identical  to current relationships of NodeJS vs W3C JS vs Deno vs Bun.

**Comment by bulk88** [comment]
About if to write a IO plugin at the PerlIO vtable level, or the iperlsys.h level. I would recommend doing it at the iperlsys.h level. 40%-50% of PP stdio keywords and operators, do not have vtable entries in the PerlIO API, and that 40-50% of PP opcodes directly call the PerlioL_*() macros that are the front end of iperlsys.h API.

Another nasty design problem is plenty of Perl C code and PP code, assume fileno() and fopenf() magically always work on all PP high level blessed $IO/*IO objects. 

How on earth do you get a Linux kernel recognized "fileno()" number for a tied $scalar_string PerlIO object? 🤔 

It becomes a worse mess on Windows, where 1 LibC of multiple LibCs has to participate with POSIX emulation of fileno() numbers that count upwards from index 2, and then map those emulated POSIX FD numbers to real Win32 API file handles, which are reference counted NT kernel objects which abstract numbers starting around 0x50 and going upwards to around 0x200s-0x800s.

NT kernel handles are 32 bit byte offsets into an array of 32 bit pointers on 32b and 64b CPUs. They always increment in units of 0x4. MS devs in some places as undocumented behavior have stolen the lower 2 bits of NT kernel handles , to use them for various purposes in the front end Win32 public API layer. 

NT kernel always masks off and ignores the lowest 2 bits of any ref counted kernel handles passed to it. 

Some examples of lower 2 bits stealing is all console I/O handles (cmd.exe stdin stdout stderr) are unaligned, b/c the lower 2 bits aren't zero, and that tells all the Win32 front Public API IO calls like ReadFile/WriteFile, NOT to send that function call to the Ring 0 NT Kernel, and instead use the seceet "mystery meat special" cmd.exe I/O API which is NOT a block/stream device based API and NOWHERE near the traditional concepts of terminal IO that we all know. 

Cmd.exe/TUI/STD* apps communicate through a MS non-public API fixed length packetized message queue/event queue system, along with small inter process shared virtual memory ring buffers, and a special malloc/realloc/free library that is synchronized/atomic safe between 2 processes on 2 cpu cores, through a shared virtual memory window.

I belive due to none stop end developer pressure, in Win 11, stdin/stdout/stderr are now real legit NT Kernel I/O handles, just like disk files, named pipes, and tcpip sockets are. I don't know if the new Win 11 kernel recognized Console IO handles are a low effort/minimum effort implementation, or a high performance implementation  internally. 

I know MS devs had the option of implementing that feature the quick cheap and fast to write the code way, by writing a simple and slow user mode Ring 3 I/O driver to manage console I/O handles. The (very painful to use) APIs have existed for decades in Windows OS to do that. For example Winsock library user mode middle ware and user mode backend and kernel mode backend, allow any end user to write a fake Ethernet Adapter or Dialup modem that lives inside a single threaded end user written user mode .dll.. The end user is required to implement all the typical BSD sockets API methods and IOCTLs codes all apps and all devs expect in the vtable they present to the winsock middle ware. And winsock then presents that fake tcpip adapter with a fake BSD sockets viable, to all other processes as the default TCP/IP adapter on that system. Most 1990s VPNs, Firewalls, and AOL SW, used this fake NIC in Ring 3 user mode API.

Winsocks's Ring 0 driver AFD.sys, will hand out real NT kernel I/O handles to all processes for this fake TCP/IP nick, and then serialized/packetize the "events" back to whatever Ring 3 process implements that fake TCP/IP NIC. 

I belive Win10 deprecated and Win 11 removed this   ancient not-a-feature/poorly designed user mode ring 3 fake sockets/NICs API finally.  Nobody, including me, will miss it. That API had a nice 30 year lifespan. And p5p.git/win32/*.c has 1-3 screen full s of code to deal with various edge case behavior of these Ring 3 user mode in a another process socket file descriptor objects.

I belive there has been a public perl users demand for Win Perl to keep that code around in the mid 2010s/late 2010s , not sure if there r still requests to keep it in the early 2020s. IIRC there are large enterprise perl users with some ancient Fortinet/Cisco VPN /faux VLAN, Radius Hardware protected Ethernet jacks on their corporate campus. So WinPerl has to have basically TCPIP functionality with that legacy Win32 VPN software with Ring 3 user mode bsd socket drivers, or Win Perl is unusuable on that corporate campus.

Perl 5's fileno() high level exposure is the biggest problem of writing a new PerlIO or iperlsys.h backend. I belive Ruby and Perl are the only 2 modern non-C programming languages that allow end devs to see/know of the POSIX FD numbers.

Python USED to allow pypy level knowledge of POSIX fileno() integers but it was deprecated and removed 20 years ago in a Python PEP and is never coming back.

For your .zip /.bz2 self decompressing .pl apps fatpacker code, I don't think perl's high level source code filter API or the INC hooks will work for what you want. My guessing is using those APIs severely corrupts caller()/warn/__PACKAGE__ logic. It can be partially repaired with perl 5 grammars rarely seen or used # line or # file keywords, but when I've tried to use it, I forgot what exactly went wrong, but # file / # line don't work inside a begin block, and  / OR caller(2) didn't see those # file # line directives from inside nested BEGIN{} BLOCKS. The directives do work in RUN phase after BEGIN / CHECK phase, but not during BEGIN/CHECK phase.

also hooking keyword require will have a lot of problems,  if your fake file names can't actually be stat()Ed or -X Ed by other perl core C code, or by Dynaloader and XSLoader. AND stat() /-X  don't appear on the PerlIO vtable method list.

There fore hooking way lower level, on the libc/almost kernel syscall IO level with iperlhost.h is the correct API is work with.

**Comment by vadimkantorov** [comment]
My attempts at libc/syscall function overrides for Perl module loading from virtual FS succeeded. But could probably be better to also be able to define these overrides via redefining some Perl IO macros - and have an official example of doing so. This would also allow to not meddle with linker...

Or even have an official example of doing libc IO function overrides with linker (command line argument `--wrap`)

Essentially, having hooks to load `*.pm` modules from memory (or another storage) is important for the embedding story
